apisec/internal/timed/set.go

```diff
@@ -0,0 +1,207 @@
+	// rebuilding is a flag to indicate whether the table is being rebuilt as
+	// part of an eviction request.
+	rebuilding atomic.Bool
+}
+
+// NewSet initializes a new, empty [Set] with the given interval and clock
+// function. The provided interval must be at least 1 second, and may not exceed
+// [config.Interval].
+//
+// Note: timestamps are stored at second resolution, so the interval will be
+// rounded down to the nearest second.
+func NewSet(interval time.Duration, clock ClockFunc) *Set {
+	if interval < time.Second {
+		panic(fmt.Errorf("NewSet: interval must be at least 1s, got %v", interval))
+	}
+	if interval > config.Interval {
+		panic(fmt.Errorf("NewSet: interval must not exceed %s, got %v", config.Interval, interval))
+	}
+
+	intervalSeconds := uint32(interval.Seconds())
+	set := &Set{
+		clock:           newBiasedClock(clock, intervalSeconds),
+		intervalSeconds: intervalSeconds,
+		zeroKey:         rand.Uint64(),
+	}
+
+	// That value cannot be zero...
+	for set.zeroKey == 0 {
+		set.zeroKey = rand.Uint64()
+	}
+
+	set.table.Store(&table{})
+
+	return set
+}
+
+// Hit determines whether the given key should be kept or dropped based on the
+// last time it was sampled. If the table grows larger than [maxItemCount], the
+// [timedSet.rebuild] method is called in a separate goroutine to begin the
+// eviction process. Until this has completed, all updates to the [timedSet] are
+// effectively dropped, as they happen on the soon-to-be-replaced table.
+//
+// Note: in order to run completely lock-less, [Set] cannot store the 0 key in
+// the table, as a 0 key is used as a sentinel value to identify free entries.
+// To avoid this pitfall, [Set.zeroKey] is used as a substitute for 0, meaning
+// 0 and [Set.zeroKey] are treated as the same key. This is not an issue in
+// common use, as given a uniform distribution of keys this only happens 1 in
+// 2^64-1 times.
+func (m *Set) Hit(key uint64) bool {
+	if key == 0 {
+		// The 0 key is used as a way to imply a slot is empty; so we cannot store
+		// it in the table. To address this, when passed a 0 key, we will use the
+		// [timedSet.zeroKey] as a substitute.
+		key = m.zeroKey
+	}
+
+	now := m.clock.Now()
+	threshold := now - m.intervalSeconds
+
+	var (
+		table = m.table.Load()
+		entry *entry
+	)
+	for {
+		var exists bool
+		entry, exists = table.FindEntry(key)
+		if exists {
+			// The entry already exists, so we can proceed...
+			break
+		}
+
+		// We're adding a new entry to the table, so we need to:
+		// 1. Ensure we have capacity (possibly trigger an eviction rebuild)
+		// 2. Claim the slot (or look for another slot if it's already claimed)
+		newCount := table.count.Add(1)
+		if newCount > config.MaxItemCount && m.rebuilding.CompareAndSwap(false, true) {
+			// We're already holding the maximium number of items, so we will rebuild
+			// in order to perform an eviction pass. Updates made in the meantime will
+			// be lost.
+			go m.rebuild(table, threshold)
```

Comment:
are we sure about this unbounded goroutine creation?

Reply:
This can only happen if `m.rebuilding.CompareAndSwap(false, true)` swaps... so there can be only 1 of these at a time.